File: python_files/raylib/bouncing_balls_v2.py
import pyray as p
from random import uniform
from math import radians, cos, sin

SCREEN_WIDTH: int = 800
SCREEN_HEIGHT: int = 600
SCREEN_TITLE: str = "Bouncing Balls V2"
SCREEN_BACKGROUND: p.Color = p.RAYWHITE

NUM_OF_BALLS: int = 200


class Ball:
    def __init__(self) -> None:
        self.radius: float = uniform(5, 30)
        self.pos: p.Vector2 = p.Vector2(
            uniform(self.radius, p.get_screen_width() - self.radius),
            uniform(self.radius, p.get_screen_height() - self.radius),
        )
        speed: float = uniform(100, 300)
        angle_radians: float = radians(uniform(0, 360))
        self.velocity: p.Vector2 = p.Vector2(
            cos(angle_radians) * speed, sin(angle_radians) * speed
        )
        self.color: p.Color = p.Color(
            p.get_random_value(0, 255),
            p.get_random_value(0, 255),
            p.get_random_value(0, 255),
            255,
        )

    def draw(self) -> None:
        p.draw_circle_v(self.pos, self.radius, self.color)

    def update(self) -> None:
        self.pos.x += self.velocity.x * p.get_frame_time()
        self.pos.y += self.velocity.y * p.get_frame_time()

        if self.pos.x < self.radius or self.pos.x > p.get_screen_width() - self.radius:
            self.velocity.x *= -1
        if self.pos.y < self.radius or self.pos.y > p.get_screen_height() - self.radius:
            self.velocity.y *= -1


def main() -> None:
    p.init_window(SCREEN_WIDTH, SCREEN_HEIGHT, SCREEN_TITLE)

    balls: list[Ball] = []
    for i in range(NUM_OF_BALLS):
        balls.append(Ball())

    while not p.window_should_close():
        for ball in balls:
            ball.update()

        p.begin_drawing()
        p.clear_background(SCREEN_BACKGROUND)

        for ball in balls:
            ball.draw()

        p.end_drawing()

    p.close_window()


if __name__ == "__main__":
    main()

File: cpp_files/raylib/bouncing_balls_v2.cpp
#include <raylib.h>
#include <raymath.h>
#include <cmath>

#define NUM_OF_BALLS 200

class Ball
{
  public:
    Vector2 pos;
    float radius;
    Vector2 velocity;
    Color color;

    Ball()
    {
        radius = GetRandomValue(10, 30);
        pos.x = GetRandomValue(radius, GetScreenWidth() - radius);
        pos.y = GetRandomValue(radius, GetScreenHeight() - radius);
        float speed = GetRandomValue(100, 300);
        float angle = GetRandomValue(0, 360) * DEG2RAD;
        velocity.x = cos(angle) * speed;
        velocity.y = sin(angle) * speed;
        color = (Color){ static_cast<unsigned char>(GetRandomValue(0, 255)),
                         static_cast<unsigned char>(GetRandomValue(0, 255)),
                         static_cast<unsigned char>(GetRandomValue(0, 255)),
                         255 };
    }

    void draw() const
    {
        DrawCircleV(pos, radius, color);
    }

    void update()
    {
        // move ball
        pos.x += velocity.x * GetFrameTime();
        pos.y += velocity.y * GetFrameTime();

        // ball bounds
        if (pos.x < radius || pos.x > GetScreenWidth() - radius)
        {
            velocity.x *= -1;
        }
        if (pos.y < radius || pos.y > GetScreenHeight() - radius)
        {
            velocity.y *= -1;
        }
    }
};

int
main(void)
{
    const int SCREEN_WIDTH = 800;
    const int SCREEN_HEIGHT = 600;
    const char SCREEN_TITLE[] = "Bouncing Balls";
    const Color SCREEN_BACKGROUND = LIGHTGRAY;

    InitWindow(SCREEN_WIDTH, SCREEN_HEIGHT, SCREEN_TITLE);

    Ball balls[NUM_OF_BALLS];

    for (int i = 0; i < NUM_OF_BALLS; i++)
    {
        Ball ball;

        balls[i] = ball;
    }

    while (!WindowShouldClose())
    {
        for (int i = 0; i < NUM_OF_BALLS; i++)
        {
            balls[i].update();
        }

        BeginDrawing();
        ClearBackground(SCREEN_BACKGROUND);

        for (int i = 0; i < NUM_OF_BALLS; i++)
        {
            balls[i].draw();
        }

        EndDrawing();
    }

    CloseWindow();

    return 0;
}